k2/csrc/fsa_utils_inl.h

```diff
@@ -0,0 +1,238 @@
+    int32_t num_rows =
+        row_end - row_begin;  // num_rows is num_states in this batch
+    K2_CHECK_LT(num_rows, arc_sum_row_splits.Dim());
+    // we always use the first num_rows elements in arc_sum_row_splits.
+    Array1<int32_t> sum_sub_range =
+        arc_sum_row_splits.Range(0, num_rows + 1);  // +1 for the last element
+    {
+      ParallelRunner pr(c);
+      // get entering arc scores
+      {
+        cudaStream_t stream = pr.NewStream();
+        With w(stream);
+        auto lambda_set_entering_arc_score = [=] __host__ __device__(
+                                                 int32_t idx3) {
+          // all idx** in below code are the indexes to entering_arc_batches
+          int32_t idx0123 = entering_arc_start_index_data[i] + idx3;
+          int32_t idx012 = entering_arc_row_ids[idx0123];
+          int32_t idx01 = state_in_batch_row_ids[idx012];
+          K2_CHECK_EQ(idx01 / num_fsas, i);  // idx01/num_batches is batch_id
+          int32_t fsa_id = idx01 % num_fsas;
+
+          int32_t entering_arc_id = entering_arc_ids[idx0123];
+          float curr_arc_score = arcs[entering_arc_id].score;
+          int32_t src_state_idx1 = arcs[entering_arc_id].src_state;
+          int32_t src_state_idx01 = fsa_row_splits1[fsa_id] + src_state_idx1;
+          arc_scores_data[idx0123] =
+              state_scores_data[src_state_idx01] + curr_arc_score;
+        };
+        Eval(stream, cpu_entering_arc_nums[i], lambda_set_entering_arc_score);
+      }
+      {
+        cudaStream_t stream = pr.NewStream();
+        With w(stream);
+        // make entering arc row splits info in each batch starting from zero,
+        // we will use it to call MaxPerSublist or LogSumPerSubList
+        int32_t *sum_splits_data = sum_sub_range.Data();
+        auto lambda_set_row_splits_for_sum =
+            [=] __host__ __device__(int32_t idx) {
+              sum_splits_data[idx] = entering_arc_row_splits[idx + row_begin] -
+                                     entering_arc_row_splits[row_begin];
+            };
+        Eval(stream, num_rows + 1, lambda_set_row_splits_for_sum);
+      }
+    }
+    int32_t arc_start = cpu_entering_arc_start[i];
+    int32_t num_arcs_this_batch = cpu_entering_arc_nums[i];
+    Array1<FloatType> sub_scores_values =
+        entering_arc_score_values.Range(arc_start, num_arcs_this_batch);
+    RaggedShape sub_scores_shape =
+        RaggedShape2(&sum_sub_range, nullptr, sub_scores_values.Dim());
+    Ragged<FloatType> sub_scores(sub_scores_shape, sub_scores_values);
+    // we always use the first num_rows elements in score_cache.
+    Array1<FloatType> sub_state_scores = score_cache.Range(0, num_rows);
+    // get scores per state in this batch
+    MaxPerSublist(sub_scores, negative_infinity, &sub_state_scores);
+    // TODO(haowen): fix compilation issu of kMinLogDiffFloat/Double in utils.h
+    /*
+  if (log_semiring)
+    LogSumPerSublist(sub_scores, negative_infinity, &sub_state_scores);
```

Comment:
There is a compilation failure here as `kMinLogDiffFloat` is a host value, but we want to access it in kernel. Which approach do you prefer to fix this issue?  just use `#define kMinLogDiffFloat the_corresponding_value`, or with different definition for host and device (e.g, for device, use `__device__` or `__constant__` and initalized in a function which called only once when the program starts?)
https://github.com/k2-fsa/k2/blob/c2c26ab54ae5c7fba08736cd78f9fe9874c67f86/k2/csrc/utils.h#L583
https://github.com/k2-fsa/k2/blob/c2c26ab54ae5c7fba08736cd78f9fe9874c67f86/k2/csrc/utils.h#L612-L633

Reply:
> Which approach do you prefer to fix this issue?

Neither.

The approach that I would propose is to remove `static`.

```cpp
const float kMinLogDiffFloat = logf(FLT_EPSILON);   // negative! 
```

Reply:
Sorry, don't understand it, you mean removing `static` then kernel can access const variable of host directly?

Reply:
Looks like my approach does not work.

Reply:
Yes, I guess so, that's why I said I don't understand it...

Reply:
Can we borrow the implementation from GTN which does not require `kMinLogDiffFloat`:

https://github.com/facebookresearch/gtn/blob/a810690546df258b6df157298ac68bf1178d1439/gtn/functions/shortest.cpp#L20

```cpp
constexpr float kInf = std::numeric_limits<float>::infinity();
constexpr float kNegInf = -std::numeric_limits<float>::infinity();

inline float logadd(float a, float b) {
  if (a == kNegInf) {
    return b;
  }
  if (b == kNegInf) {
    return a;
  }
  return std::max(a, b) + std::log1p(std::exp(-std::abs(a - b)));
}
```

We don't need `kNegInf`. CUDA provides `isinf`:
https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__SINGLE.html#group__CUDA__MATH__SINGLE_1g0a62e45f335a23ee64ecad3fb87a72e3

> __device__​ __RETURN_TYPE isinf ( float  a )

There is also a function `isinf()` on the host side.

Reply:
I think it's fine, let's wait for @danpovey 's suggestion.

Reply:
I prefer #define, you may have to figure out the numerical value.
Comparisons with infinity do not work when you add -Ofast: they are turned off, so that's a bit dangerous

Reply:
Ok, will do `#define` in the next PR.